crypto/kyber/p_kyber_test.cc

```diff
@@ -4,280 +4,463 @@
 
-  EXPECT_EQ(0, EVP_PKEY_cmp(kyber_pkey1, kyber_pkey2));
-  EXPECT_EQ(1, EVP_PKEY_cmp(kyber_pkey1, kyber_pkey1));
-  EXPECT_EQ(1, EVP_PKEY_cmp(kyber_pkey2, kyber_pkey2));
+    EXPECT_EQ(0, EVP_PKEY_cmp(kyber_pkey1, kyber_pkey2));
+    EXPECT_EQ(1, EVP_PKEY_cmp(kyber_pkey1, kyber_pkey1));
+    EXPECT_EQ(1, EVP_PKEY_cmp(kyber_pkey2, kyber_pkey2));
 
-  EVP_PKEY_CTX_free(kyber_pkey_ctx1);
-  EVP_PKEY_CTX_free(kyber_pkey_ctx2);
+    EVP_PKEY_CTX_free(kyber_pkey_ctx1);
+    EVP_PKEY_CTX_free(kyber_pkey_ctx2);
+  }
 }
 
-TEST(Kyber512Test, NewKeyFromBytes) {
-  // Source key
-  EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KYBER512, nullptr);
-  ASSERT_NE(kyber_pkey_ctx, nullptr);
-
-  EVP_PKEY *kyber_pkey = EVP_PKEY_new();
-  ASSERT_NE(kyber_pkey, nullptr);
-
-  EXPECT_TRUE(EVP_PKEY_keygen_init(kyber_pkey_ctx));
-  EXPECT_TRUE(EVP_PKEY_keygen(kyber_pkey_ctx, &kyber_pkey));
-  ASSERT_NE(kyber_pkey->pkey.ptr, nullptr);
-  const KYBER512_KEY *kyber512Key = (KYBER512_KEY *)(kyber_pkey->pkey.ptr);
-
-  // New raw public key
-  EVP_PKEY *new_public = EVP_PKEY_new_raw_public_key(EVP_PKEY_KYBER512,
-                                                     NULL,
-                                                     kyber512Key->pub,
-                                                     KYBER512_PUBLIC_KEY_BYTES);
-  ASSERT_NE(new_public, nullptr);
-
-  uint8_t *buf = nullptr;
-  size_t buf_size;
-  EXPECT_FALSE(EVP_PKEY_get_raw_private_key(new_public, buf, &buf_size));
-  uint32_t err = ERR_get_error();
-  EXPECT_EQ(ERR_LIB_EVP, ERR_GET_LIB(err));
-  EXPECT_EQ(EVP_R_NOT_A_PRIVATE_KEY, ERR_GET_REASON(err));
-
-  // EVP_PKEY_cmp just compares the public keys so this should return 1
-  EXPECT_EQ(1, EVP_PKEY_cmp(kyber_pkey, new_public));
-
-  // New raw private key
-  EVP_PKEY *new_private = EVP_PKEY_new_raw_private_key(EVP_PKEY_KYBER512,
-                                                       NULL,
-                                                       kyber512Key->priv,
-                                                       KYBER512_SECRET_KEY_BYTES);
-  ASSERT_NE(new_private, nullptr);
-  const KYBER512_KEY *newKyber512Key = (KYBER512_KEY *)(new_private->pkey.ptr);
-  EXPECT_EQ(0, OPENSSL_memcmp(kyber512Key->priv, newKyber512Key->priv, KYBER512_SECRET_KEY_BYTES));
-
-  EVP_PKEY_CTX_free(kyber_pkey_ctx);
-  EVP_PKEY_free(new_public);
-  EVP_PKEY_free(new_private);
+TEST(KyberTest, NewKeyFromBytes) {
+  for (auto pkey_id : pkeyIDsToTest) {
+    // Source key
+    EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(pkey_id, nullptr);
+    ASSERT_NE(kyber_pkey_ctx, nullptr);
+
+    EVP_PKEY *kyber_pkey = EVP_PKEY_new();
+    ASSERT_NE(kyber_pkey, nullptr);
+
+    EXPECT_TRUE(EVP_PKEY_keygen_init(kyber_pkey_ctx));
+    EXPECT_TRUE(EVP_PKEY_keygen(kyber_pkey_ctx, &kyber_pkey));
+    ASSERT_NE(kyber_pkey->pkey.ptr, nullptr);
+
+    uint8_t *pub  = get_pub_from_kyber_key(kyber_pkey);
+    uint8_t *priv = get_priv_from_kyber_key(kyber_pkey);
```

Comment:
Does this need to be the exact same memory address or can it use EVP_PKEY_get_raw_public/private_key to get a copy and get rid of these functions?